### ISSUE TITLE
example.c:66:1: error: unknown type name ‘JSAMPLE’

### ISSUE THREAD

**Issue by kirantpatil** [libjpeg-turbo/libjpeg-turbo#178]
I am trying to compile [example.c](https://raw.githubusercontent.com/libjpeg-turbo/libjpeg-turbo/master/example.c) code using gcc and it is giving below error.

I have installed libjpeg-turbo package on ubuntu.

In example.c file I did below changes.

/*
#include "jpeglib.h"
*/

 #include "turbojpeg.h"

$ gcc example.c 
example.c:66:1: error: unknown type name ‘JSAMPLE’
 extern JSAMPLE * image_buffer; /* Points to large array of R,G,B-order data */
 ^
example.c: In function ‘GLOBAL’:
example.c:77:1: error: expected declaration specifiers before ‘write_JPEG_file’
 write_JPEG_file (char * filename, int quality)
 ^
example.c:256:25: error: field ‘pub’ has incomplete type
   struct jpeg_error_mgr pub; /* "public" fields */
                         ^
example.c:255:1: warning: empty declaration [enabled by default]
 struct my_error_mgr {
 ^
example.c:261:31: error: storage class specified for parameter ‘my_error_ptr’
 typedef struct my_error_mgr * my_error_ptr;
                               ^
example.c:267:1: error: expected declaration specifiers before ‘METHODDEF’
 METHODDEF(void)
 ^
example.c:288:1: error: expected declaration specifiers before ‘GLOBAL’
 GLOBAL(int)
 ^
example.c:76:1: error: old-style parameter declarations in prototyped function definition
 GLOBAL(void)
 ^
example.c:412:1: error: expected ‘{’ at end of input
 }
 ^

**Comment by kirantpatil** [comment]
If I keep original header then I get below error.

#include "jpeglib.h"

/*
 #include "turbojpeg.h"
*/

$ gcc example.c 
example.c: In function ‘write_JPEG_file’:
example.c:120:5: warning: incompatible implicit declaration of built-in function ‘exit’ [enabled by default]
     exit(1); /* exit(1) was there*/
     ^
/usr/bin/ld: /usr/lib/debug/usr/lib/x86_64-linux-gnu/crt1.o(.debug_info): relocation 0 has invalid symbol index 11
/usr/bin/ld: /usr/lib/debug/usr/lib/x86_64-linux-gnu/crt1.o(.debug_info): relocation 1 has invalid symbol index 12
/usr/bin/ld: /usr/lib/debug/usr/lib/x86_64-linux-gnu/crt1.o(.debug_info): relocation 2 has invalid symbol index 2
/usr/bin/ld: /usr/lib/debug/usr/lib/x86_64-linux-gnu/crt1.o(.debug_info): relocation 3 has invalid symbol index 2
/usr/bin/ld: /usr/lib/debug/usr/lib/x86_64-linux-gnu/crt1.o(.debug_info): relocation 4 has invalid symbol index 11
/usr/bin/ld: /usr/lib/debug/usr/lib/x86_64-linux-gnu/crt1.o(.debug_info): relocation 5 has invalid symbol index 13
/usr/bin/ld: /usr/lib/debug/usr/lib/x86_64-linux-gnu/crt1.o(.debug_info): relocation 6 has invalid symbol index 13
/usr/bin/ld: /usr/lib/debug/usr/lib/x86_64-linux-gnu/crt1.o(.debug_info): relocation 7 has invalid symbol index 13
/usr/bin/ld: /usr/lib/debug/usr/lib/x86_64-linux-gnu/crt1.o(.debug_info): relocation 8 has invalid symbol index 12
/usr/bin/ld: /usr/lib/debug/usr/lib/x86_64-linux-gnu/crt1.o(.debug_info): relocation 9 has invalid symbol index 13
/usr/bin/ld: /usr/lib/debug/usr/lib/x86_64-linux-gnu/crt1.o(.debug_info): relocation 10 has invalid symbol index 13
/usr/bin/ld: /usr/lib/debug/usr/lib/x86_64-linux-gnu/crt1.o(.debug_info): relocation 11 has invalid symbol index 13
/usr/bin/ld: /usr/lib/debug/usr/lib/x86_64-linux-gnu/crt1.o(.debug_info): relocation 12 has invalid symbol index 13
/usr/bin/ld: /usr/lib/debug/usr/lib/x86_64-linux-gnu/crt1.o(.debug_info): relocation 13 has invalid symbol index 13
/usr/bin/ld: /usr/lib/debug/usr/lib/x86_64-linux-gnu/crt1.o(.debug_info): relocation 14 has invalid symbol index 13
/usr/bin/ld: /usr/lib/debug/usr/lib/x86_64-linux-gnu/crt1.o(.debug_info): relocation 15 has invalid symbol index 13
/usr/bin/ld: /usr/lib/debug/usr/lib/x86_64-linux-gnu/crt1.o(.debug_info): relocation 16 has invalid symbol index 13
/usr/bin/ld: /usr/lib/debug/usr/lib/x86_64-linux-gnu/crt1.o(.debug_info): relocation 17 has invalid symbol index 13
/usr/bin/ld: /usr/lib/debug/usr/lib/x86_64-linux-gnu/crt1.o(.debug_info): relocation 18 has invalid symbol index 13
/usr/bin/ld: /usr/lib/debug/usr/lib/x86_64-linux-gnu/crt1.o(.debug_info): relocation 19 has invalid symbol index 21
/usr/bin/ld: /usr/lib/debug/usr/lib/x86_64-linux-gnu/crt1.o(.debug_line): relocation 0 has invalid symbol index 2
/usr/lib/gcc/x86_64-linux-gnu/4.8/../../../x86_64-linux-gnu/crt1.o: In function `_start':
(.text+0x20): undefined reference to `main'
/tmp/ccu83gfX.o: In function `write_JPEG_file':
example.c:(.text+0x32): undefined reference to `jpeg_std_error'
example.c:(.text+0x52): undefined reference to `jpeg_CreateCompress'
example.c:(.text+0xba): undefined reference to `jpeg_stdio_dest'
example.c:(.text+0xc0): undefined reference to `image_width'
example.c:(.text+0xcc): undefined reference to `image_height'
example.c:(.text+0xf5): undefined reference to `jpeg_set_defaults'
example.c:(.text+0x111): undefined reference to `jpeg_set_quality'
example.c:(.text+0x125): undefined reference to `jpeg_start_compress'
example.c:(.text+0x12b): undefined reference to `image_width'
example.c:(.text+0x140): undefined reference to `image_buffer'
example.c:(.text+0x179): undefined reference to `jpeg_write_scanlines'
example.c:(.text+0x198): undefined reference to `jpeg_finish_compress'
example.c:(.text+0x1b6): undefined reference to `jpeg_destroy_compress'
/tmp/ccu83gfX.o: In function `read_JPEG_file':
example.c:(.text+0x28e): undefined reference to `jpeg_std_error'
example.c:(.text+0x2c8): undefined reference to `jpeg_destroy_decompress'
example.c:(.text+0x2fa): undefined reference to `jpeg_CreateDecompress'
example.c:(.text+0x313): undefined reference to `jpeg_stdio_src'
example.c:(.text+0x327): undefined reference to `jpeg_read_header'
example.c:(.text+0x336): undefined reference to `jpeg_start_decompress'
example.c:(.text+0x396): undefined reference to `jpeg_read_scanlines'
example.c:(.text+0x3b5): undefined reference to `put_scanline_someplace'
example.c:(.text+0x3d4): undefined reference to `jpeg_finish_decompress'
example.c:(.text+0x3e3): undefined reference to `jpeg_destroy_decompress'
collect2: error: ld returned 1 exit status
fabritronix@fabritronix:~/Documents/lib_jpeg_turbo_example$

**Comment by dcommander** [comment]
example.c is part of the documentation.  It is, as described in libjpeg.txt, "heavily commented skeleton code for calling the JPEG library."  It is not meant to be compiled as a standalone program, since it has no main() function.  See cjpeg.c and djpeg.c and jpegtran.c for examples of libjpeg API usage that can be (and are) compiled as standalone programs.

The architecture, documentation, and general source code layout of the libjpeg API portions of libjpeg-turbo were inherited from libjpeg.  example.c dates from 1994 and has changed very little in the 23 years since.  Our project was originally created as a SIMD-accelerated fork of libjpeg v6b, which itself is nearly 20 years old.  I have taken pains to get rid of a lot of unnecessary legacy code, including MS-DOS support and other stuff that is not needed on modern systems, but I've also taken pains to retain backward compatibility with libjpeg v6b as much as possible, since it was a de facto industry standard (and our project has more or less inherited that role.)

This is largely a single-person project, and I don't have time to give the millions of libjpeg-turbo users in the world personal assistance with the library.  Feel free to post community support requests to the libjpeg-turbo-users mailing list, a link to which can be found on libjpeg-turbo.org.  GitHub issues are mostly just going to be read by me, so I would appreciate it if you would confine these issues to actual bug reports and feature requests.

**Comment by mlfarrell** [comment]
I'm having the exact same error.  Your build system has absolutely zero documentation on how to extract headers for use after a build.  I'm completely lost here.  Please open this back up and help us!

**Comment by dcommander** [comment]
@mlfarrell please take the time to read my comment above, which explains in plain English that example.txt is basically something between pseudo-code and an actual program. I renamed it with a .txt extension as a result of this issue, in order to make that crystal clear. If you want a more straightforward example for first-time users, an example that can actually be built and run, then use the TurboJPEG API and look at tjexample.c. Otherwise, cjpeg.c and djpeg.c are buildable and runnable examples of the libjpeg API. example.txt is just meant to show you the basic program flow.

Please understand that the libjpeg API is nearly 30 years old. The examples and documentation for that API have been improved in various ways as part of this project, but they largely function the way they did in the 1990s. libjpeg-turbo was originally just that: a turbo implementation of libjpeg. After 10 years, it’s now much more than that, and that’s in part because I have spent the equivalent of hundreds of thousands of dollars in unpaid labor improving this code base over the last ten years. And yet people still complain ... I’m sorry if you’re having trouble compiling a text file that wasn’t meant to be compiled, but that is so not my problem.

**Comment by mlfarrell** [comment]
I understand your frustration. You have to understand that I’ve been equally frustrated trying to get libJPEG working on android  spending several hours trying to do so. I tracked the above build issue down to the fact that the jconfig headers have not been properly generated aka they do not contain a definition for JSAMPLE size (or similarly named)  Out of frustration, I manually it inserted it into the config files and assumed a value of 8 from what I can tell from the CMake Lists.

**Comment by mlfarrell** [comment]
Also accept my apology for being a dick.  I let my frustration get the best of me. I understand how difficult it is to maintain open source software. It’s a thankless task. Best of luck.

**Comment by dcommander** [comment]
@mlfarrell That is in no way related to this issue.  This specific issue was due to a misunderstanding on the part of the OP regarding the fact that the libjpeg API example file is not a self-contained C program.  You asked me to re-open the issue, which implied that you were having the same issue, but you weren't.  That indicates that you didn't actually read anything above your comment.  The jconfig headers are generated if you follow the Android build instructions in BUILDING.md.  If you don't follow the Android build instructions, then you're on your own.

**Comment by mlfarrell** [comment]
You're being overly rude after I went out of my way to apologize.  Either way turbo jpeg never worked for me on android, it failed to open basic jpeg files.  I went with another implementation of libjpeg which not only built properly for android, also properly generated the jconfig and jmoreconfig headers.  Make of that what you will. Peace.

**Comment by dcommander** [comment]
In what way am I being rude?  I'm just explaining why your issue was unrelated to this issue and giving you advice as to how to fix the Android build problem you are having.  I am an independent OSS developer, and I rely on these GitHub projects in order to make money, which is necessary in order for me to continue developing these projects.  Thus, when you comment on an issue here, you are entering my place of business.  I read and appreciated your apology, but I deleted it because it contained unprofessional language (swear words), which is not appropriate in my place of business.  You went on to explain how you were frustrated by the Android build not working, and I felt it necessary to explain that that is in no way what this issue was ever about.  Nothing above makes you sound like a jerk except your most recent comment, and that is entirely on you, not me.  If you are still having Android build issues after following the instructions, then post a new issue about that, and I'll look into it.  But, again, this thread was never about Android build issues, so your comments never belonged here in the first place.

**Comment by dcommander** [comment]
BTW, how you extract headers for use after a build is by installing the code, via `make install` (in the case of an Android build.)  That is true for every open source library that has generated headers, and I haven't seen any other libraries that explicitly document that.
